crates/supervisor/core/src/rewinder/chain.rs

```diff
@@ -0,0 +1,94 @@
+use alloy_primitives::ChainId;
+use derive_more::Constructor;
+use kona_interop::DerivedRefPair;
+use kona_supervisor_storage::{LogStorageReader, StorageError, StorageRewinder};
+use thiserror::Error;
+use tracing::{error, info, warn};
+
+/// Initiates supervisor-level rewinds based on chain events or storage conflicts.
+///
+/// This coordinates per-chain rewind logic using the underlying [`StorageRewinder`] implementation,
+/// and encapsulates the context in which rewinds should occur.
+///
+/// It is used in response to:
+/// - Local derivation conflicts (failure updating supervisor state)
+/// - L1-originated reorgs affecting derived state
+#[derive(Debug, Constructor)]
+pub struct ChainRewinder<DB> {
+    chain_id: ChainId,
+    db: DB,
+}
+
+#[allow(dead_code)] // todo: to be removed in the follow up PR
+impl<DB> ChainRewinder<DB>
+where
+    DB: StorageRewinder + LogStorageReader,
+{
+    /// Handles a local reorg by rewinding supervisor state from the conflicting derived pair.
+    ///
+    /// This is triggered when an update to supervisor storage fails due to an
+    /// integrity violation (e.g., mismatched on storing local safe block hash).
+    fn handle_local_reorg(&self, derived_pair: &DerivedRefPair) -> Result<(), StorageError> {
+        warn!(
+            target: "rewinder",
+            chain = %self.chain_id,
+            derived_block = %derived_pair.derived,
+            "Local derivation conflict detected — rewinding..."
+        );
+        // get the same block from log storage
+        let conflicting_block =
+            self.db.get_block(derived_pair.derived.number).inspect_err(|err| {
+                error!(
+                    target: "rewinder",
+                    chain = %self.chain_id,
+                    block_number = derived_pair.derived.number,
+                    %err,
+                    "Error retrieving conflicting block for reorg"
+                );
+            })?;
+
```

Comment:
I think we should cross check if the block is conflicting

Reply:
added the check.